### PR TITLE
Add select service namespace in catalog request form

### PR DESCRIPTION
**Pull request by jkupferer**
- Project namespace select only appears if multiple namespaces are available to user.
- Move ServiceNamespaceSelect to components
- Change ServiceNamsepaceSelect to patternfly ContextSelector
- Show workshops in navigation to all users

### CODE REVIEW

File: catalog/ui/src/app/Catalog/CatalogItemForm.tsx [hunk 1]
```diff
@@ -79,20 +80,46 @@ const ScheduleModal: React.FC<{ defaultTimestamp: number; onSelect: (date: Date)
...
-  const { isAdmin, groups, roles, workshopNamespaces, userNamespace } = useSession().getSession();
+  const { isAdmin, groups, roles, serviceNamespaces: sessionServiceNamespaces, workshopNamespaces: sessionWorkshopNamespaces, userNamespace } = useSession().getSession();
+  const enableFetchUserNamespaces: boolean = isAdmin;
   const { userImpersonated } = useImpersonateUser();
   const { data: catalogItem } = useSWR<CatalogItem>(
-    apiPaths.CATALOG_ITEM({ namespace, name: catalogItemName }),
+    apiPaths.CATALOG_ITEM({ namespace: catalogNamespaceName, name: catalogItemName }),
     fetcher
   );
+  const [namespaceSelectIsOpen, setNamespaceSelectIsOpen] = useState(false);
```
Comment (aleixhub) [inline catalog/ui/src/app/Catalog/CatalogItemForm.tsx:97]: I can't see that this is being used

File: catalog/ui/src/app/Catalog/CatalogItemForm.tsx [hunk 2]
```diff
@@ -176,10 +203,10 @@ const CatalogItemFormData: React.FC<{ namespace: string; catalogItemName: string
     } else {
       const resourceClaim = await createServiceRequest({
         catalogItem,
-        catalogNamespaceName: namespace,
-        userNamespace,
+        catalogNamespaceNameName: catalogNamespaceName,
```
Comment (aleixhub) [inline catalog/ui/src/app/Catalog/CatalogItemForm.tsx:206]: typo: catalogNamespaceName

File: catalog/ui/src/app/components/ServiceNamespaceSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import { apiPaths, fetcher } from '@app/api';
+
+import './service-namespace-select.css';
+
+const ServiceNamespaceSelect: React.FC<{
+  allowSelectAll: boolean;
+  currentNamespaceName: string;
+  isPlain: boolean;
+  isText: boolean;
+  serviceNamespaces: ServiceNamespace[];
```
Comment (aleixhub) [inline catalog/ui/src/app/components/ServiceNamespaceSelect.tsx:15]: not needed, is defined inside the component

File: catalog/ui/src/app/components/ServiceNamespaceSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  onSelect: (namespace: ServiceNamespace) => void;
+  selectWorkshopNamespace: boolean; 
+}> = ({
+  allowSelectAll,
+  currentNamespaceName,
+  isPlain,
+  isText,
```
Comment (aleixhub) [inline catalog/ui/src/app/components/ServiceNamespaceSelect.tsx:22]: add a default value and make it optional (isPlain, isText, allowSelectAll)

File: catalog/ui/src/app/Catalog/CatalogItemForm.tsx (hunk 1 above)
Comment (jkupferer) [reply catalog/ui/src/app/Catalog/CatalogItemForm.tsx]: Removed unused variables.

File: catalog/ui/src/app/Catalog/CatalogItemForm.tsx (hunk 2 above)
Comment (jkupferer) [reply catalog/ui/src/app/Catalog/CatalogItemForm.tsx]: Fixed.

File: catalog/ui/src/app/components/ServiceNamespaceSelect.tsx (hunk 3 above)
Comment (jkupferer) [reply catalog/ui/src/app/components/ServiceNamespaceSelect.tsx]: Removed.

File: catalog/ui/src/app/Catalog/CatalogItemForm.tsx [hunk 5]
```diff
@@ -79,20 +80,46 @@ const ScheduleModal: React.FC<{ defaultTimestamp: number; onSelect: (date: Date)
...
+  const enableFetchUserNamespaces: boolean = isAdmin;
   const { userImpersonated } = useImpersonateUser();
   const { data: catalogItem } = useSWR<CatalogItem>(
-    apiPaths.CATALOG_ITEM({ namespace, name: catalogItemName }),
+    apiPaths.CATALOG_ITEM({ namespace: catalogNamespaceName, name: catalogItemName }),
     fetcher
   );
+  const [namespaceSelectIsOpen, setNamespaceSelectIsOpen] = useState(false);
   const [userRegistrationSelectIsOpen, setUserRegistrationSelectIsOpen] = useState(false);
+  const { data: userNamespaceList } = useSWR<NamespaceList>(
```
[the remark this answers is not in the document]
Comment (jkupferer) [reply catalog/ui/src/app/Catalog/CatalogItemForm.tsx]: Removed fetching namespace list.

File: catalog/ui/src/app/components/ServiceNamespaceSelect.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  currentNamespaceName,
+  isPlain = false,
+  isText = false,
+  onSelect,
+  selectWorkshopNamespace=false,
+}) => {
+  const [isOpen, setIsOpen] = useState(false);
+  const [searchValue, setSearchValue] = React.useState('');
+  const { isAdmin, serviceNamespaces: sessionServiceNamespaces, workshopNamespaces: sessionWorkshopNamespaces } = useSession().getSession();
+  const enableFetchNamespaces: boolean = isAdmin;
```
Comment (aleixhub) [inline catalog/ui/src/app/components/ServiceNamespaceSelect.tsx:28]: nip: as isAdmin is known that is a boolean, typescript already infers that type, therefore is not needed to declare it again

File: catalog/ui/src/app/Catalog/CatalogItemFormReducer.ts [hunk 7]
```diff
@@ -339,12 +348,22 @@ function reduceFormStateTermsOfServiceAgreed(initialState: FormState, termsOfSer
...
-function reduceFormStateWorkshop(initialState: FormState, workshop: WorkshopProps = null): FormState {
+function reduceFormStateWorkshop(
+  initialState: FormState,
+  allowServiceNamespaces: ServiceNamespace[],
+  serviceNamespace: ServiceNamespace,
+  workshop: WorkshopProps = null
+): FormState {
   const isSalesforceIdRequired = salesforceIdRequired({ ...initialState, workshop });
+  const resetServiceNamepace = true;
+  allowServiceNamespaces && !(
```
Comment (aleixhub) [inline catalog/ui/src/app/Catalog/CatalogItemFormReducer.ts:359]: i guess that this is not correct, you might wanted to assign this variable to the boolean above?

File: catalog/ui/src/app/components/ServiceNamespaceSelect.tsx (hunk 6 above)
Comment (jkupferer) [reply catalog/ui/src/app/components/ServiceNamespaceSelect.tsx]: removed declaration of boolean type.

File: catalog/ui/src/app/Catalog/CatalogItemFormReducer.ts (hunk 7 above)
Comment (jkupferer) [reply catalog/ui/src/app/Catalog/CatalogItemFormReducer.ts]: Good catch. I had put in a change here to force the value to test and didn't fix it.

**Review by aleixhub — APPROVED** [review]
LGTM

**Pull request merged**